Zotero/Controllers/Sync/SyncController.swift

```diff
@@ -1747,12 +1771,23 @@ final class SyncController: SynchronizationController {
             self.queue.removeAll()
             self.enqueue(actions: actions, at: 0, delay: delay)
 
-        case .libraryConflict:
+        case .libraryConflict(let response):
             guard self.conflictRetries < self.conflictDelays.count else {
                 self.abort(error: .cantResolveConflict(data: .from(syncObject: object, keys: keys, libraryId: libraryId)))
                 return true
             }
 
+            if response == "If-None-Match: * set but file exists" {
```

Comment:
Don't want to test against this — regular English strings (i.e., not JSON with a specific structure) in the response body are just for debugging and aren't guaranteed to be stable. The relevant conditions here would be 1) a `412` for a file upload authorization request and 2) `Last-Modified-Version` of the response > the local attachment version. And then there are some other situations that the desktop app handles if (2) isn't true:

https://github.com/zotero/zotero/blob/6106e379c96a26897ead32a300190e2327a11f3e/chrome/content/zotero/xpcom/storage/zfs.js#L552-L620

Also doesn't seem right that this is under something called `libraryConflict`. This would specifically be for a file upload authorization request. You can see the other 4xx response codes handled by the desktop for a file upload authorization request in the function linked above. I would expect this to be handled in the same place.

Reply:
@dstillman I'm not sure I understand fully. In case of 412 on authorization it first checks response header version, compares to local item version and if remote version is newer, it re-downloads the file and updates item. 

But if remote `version <= item.version` then it fetches the same item from backend and checks versions again? Wouldn't `Last-Modified-Version` from authorization response be the same as `version` of item fetched directly from backend?

Then if remote `version <= item.version` it checks md5 hashes of remote and local files, if they match it just updates the local file as synced. If they don't match, it marks a file conflict, which we probably don't support on iOS yet. So I guess I'll just abort sync in that case?

Also for 413 I guess I'll just show an error with usage and quota values?

Reply:
Yeah, sorry, that's really confusing ("downloading" is referring to the attachment item, not the file), possibly evolved incorrectly over the years, and probably isn't relevant in any case as long as there's no local file editing.

It looks like there are three possible causes of 412 for a file upload authorization request currently:

1) If-Match being set to a hash but the file not existing
2) The ETag not matching the current version of the file
3) If-None-Match: * being set but the file existing

(1) won't ever happen from iOS, I don't think.

(2) and (3) can probably be simplified to force-downloading the file and marking as uploaded, yes, at least as long as we don't have local file editing.